jetty-util/src/main/java/org/eclipse/jetty/util/thread/strategy/AdaptiveExecutionStrategy.java

```diff
@@ -32,27 +32,44 @@
+ * of both the task and the current execution to select an optimal strategy
+ * that prioritizes executing the task immediately in the current thread
+ * if can be done so without thread starvation issues.</p>
  *
- * <p>This strategy preemptively dispatches a thread as a pending producer, so that
- * when a thread produces a task it can immediately run the task and let the pending
- * producer thread take over production. When operating in this way, the sub-strategy
- * is called Execute Produce Consume (EPC).</p>
- * <p>However, if the task produced uses the {@link Invocable} API to indicate that
- * it will not block, then the strategy will run it directly, regardless of the
- * presence of a pending producer thread and then resume production after the
- * task has completed. When operating in this pattern, the sub-strategy is called
- * ProduceConsume (PC).</p>
- * <p>If there is no pending producer thread available and if the task has not
- * indicated it is non-blocking, then this strategy will dispatch the execution of
- * the task and immediately continue production. When operating in this pattern, the
- * sub-strategy is called ProduceExecuteConsume (PEC).</p>
+ * <p>If a produced task has used the {@link Invocable} API to indicate that
+ * it is {@link Invocable.InvocationType#NON_BLOCKING},
+ * then the task will always be run directly and production resumed afterwards.
+ * When operating in this pattern, the sub-strategy is called ProduceConsume (PC).</p>
+ *
+ * <p>If the producing thread has indicated that it is itself
+ * {@link Invocable.InvocationType#NON_BLOCKING}, then produced tasks that
+ * are {@link Invocable.InvocationType#EITHER}, are invoked directly as
+ * {@link Invocable.InvocationType#NON_BLOCKING} and then production resumed.
+ * This sub-strategy is called ProduceInvokeConsume (PIC).</p>
+ *
+ * <p>For all other tasks produced by a {@link Invocable.InvocationType#NON_BLOCKING}
+ * thread, the task is dispatched for execution by a thread pool and the current
+ * producing thread immediately continues production.
+ * This sub-strategy is called ProduceExecuteConsume (PEC).</p>
+ *
+ * <p>If the producing thread may block, then for produced task that also may block,
+ * the strategy may attempts to dispatch another producing thread via a {@link TryExecutor}.
+ * If either there is already a producer thread pending or the call
+ * to {@link TryExecutor#tryExecute(Runnable)} succeeds, then the task
+ * is executed directly and the pending producer will take over production.
+ * This sub-strategy is called Execute Produce Consume (EPC), but was previously
+ * named EatWhatYouKill (EWYK) after a hunting proverb, in the sense that one
+ * should kill(produce) only to eat(consume).</p>
+ *
+ * <p>If there is no pending producer thread available and the produce task
+ * is {@link Invocable.InvocationType#EITHER}, then the PIC sub-strategy is
+ * used. Otherwise the PEC sub-strategy is used.</p>
```

Comment:
IMHO the whole javadoc can be improved in this way:

Opening paragraph is good.

Then mention that a task can use `invokeNonBlocking()`, therefore marking the _thread_ as non-blocking.
Then mention that strategies may be cascaded, so the next can know if it has been invoked from a non-blocking thread.

Then mention the various cases, using bullet lists or `dl`, `dd`, `dt` to enumerate the various cases with less prose.
Something like:

* Invoked from non-blocking thread; task produced is:
** BLOCKING: foo...
** NON_BLOCKING: bar...
** EITHER: baz...

etc.

Reply:
Agreed.   I rewrote it in that style and ultimately it came out a lot simpler.... so much so that it indicated that the actual code itself was too complex when selecting the sub-strategy.
So I rewrote the code to follow the documented substrategy choice.  I'm 99% sure it is exactly the same algorithm, just simpler.  But if there is a doubt, then let's just change documentation and leave the code.  @lachlan-roberts @lorban extra eyes would be handy.

Reply:
@gregw The algorithm looks the same to me, only difference I can see is you were previously using the lock for some of the logic and you aren't using it anymore. Was this intentional?

Reply:
Doh! Good catch!